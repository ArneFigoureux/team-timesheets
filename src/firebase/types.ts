export enum LoginProvider {
    Email = 0,
    Facebook = 1,
    Google = 2,
    Guest = 3,
}
