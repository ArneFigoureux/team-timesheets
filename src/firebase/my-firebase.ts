import "firebase/firestore";
import * as firebase from "firebase/app";
import { config } from "../config";

const app = firebase.initializeApp({
    apiKey: config.apiKey,
    authDomain: config.authDomain,
    projectId: config.projectId,
    storageBucket: config.storageBucket
});

export const firestore = app.firestore();
export const storage = app.storage();
export const auth = app.auth();
export const functions = app.functions();