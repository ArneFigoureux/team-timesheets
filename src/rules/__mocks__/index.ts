export const canEditProject = jest.fn().mockReturnValue(true);
export const canDeleteProject = jest.fn().mockReturnValue(true);
export const canManageProjects = jest.fn().mockReturnValue(true);
export const canReadUsers = jest.fn().mockReturnValue(true);
export const isAdmin = jest.fn().mockReturnValue(true);