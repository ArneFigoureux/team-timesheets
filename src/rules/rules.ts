import { IUser, IProject } from "../../common/dist";

const isAdmin = (user?: IUser) => {
    return !!(user && user.roles.admin);
}

const isEditor = (user?: IUser) => {
    return !!(user && user.roles.editor);
}

export const canAddTask = (user?: IUser) => isAdmin(user);
export const canEditTask = (user?: IUser) => isAdmin(user);
export const canDeleteTask = (user?: IUser) => isAdmin(user);

export const canAddProject = (user?: IUser) => isEditor(user);
export const canEditProject = (project: IProject, user?: IUser, userId?: string) => isAdmin(user) || (isEditor(user) && userId && userId === project.createdBy);
export const canDeleteProject = (project: IProject, user?: IUser, userId?: string) => isAdmin(user) || (isEditor(user) && userId && userId === project.createdBy);
export const canArchiveProject = (project: IProject, user?: IUser, userId?: string) => isAdmin(user) || (isEditor(user) && userId && userId === project.createdBy);

export const canAddClient = (user?: IUser) => isAdmin(user);
export const canEditClient = (user?: IUser) => isAdmin(user);
export const canDeleteClient = (user?: IUser) => isAdmin(user);

export const canManageTeams = (user?: IUser) => isAdmin(user);
export const canAddTeam = (user?: IUser) => isAdmin(user);
export const canEditTeam = (user?: IUser) => isAdmin(user);
export const canDeleteTeam = (user?: IUser) => isAdmin(user);

export const canReadUsers = (user?: IUser) => isAdmin(user);