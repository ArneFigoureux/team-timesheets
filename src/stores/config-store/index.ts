export * from "./config-store";
export * from "./use-config-store";