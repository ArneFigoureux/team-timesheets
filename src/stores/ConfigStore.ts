import { observable, computed } from 'mobx';
import { Collection, ICollection } from "firestorable";
import { IRootStore } from './RootStore';
import { IProject, ITask, IClient, ITeam } from '../../common/dist';
import { firestore } from '../firebase/myFirebase';

export interface IConfigStore {
    projects: ICollection<IProject>;
    tasks: ICollection<ITask>;
    clientsCollection: ICollection<IClient>;
    teamsCollection: ICollection<IClient>;
    clients: (IClient & {id: string})[];
    teams: (ITeam & {id: string, isSelected: boolean})[];
    taskId?: string;
    projectId?: string;
    clientId?: string;
    teamId?: string;
}

export class ConfigStore implements IConfigStore {
    //private readonly _rootStore: IRootStore;

    readonly projects: ICollection<IProject>;
    readonly tasks: ICollection<IProject>;
    readonly clientsCollection: ICollection<IClient>;
    readonly teamsCollection: ICollection<ITeam>;

    @observable.ref taskId?: string;
    @observable.ref projectId?: string;
    @observable.ref clientId?: string;
    @observable.ref teamId?: string;

    constructor(_rootStore: IRootStore, getCollection: (name: string) => firebase.firestore.CollectionReference) {
        // this._rootStore = rootStore;
        this.projects = observable(new Collection<IProject>(firestore, getCollection.bind(this, "projects"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));

        this.tasks = observable(new Collection<ITask>(firestore, getCollection.bind(this, "tasks"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));

        this.clientsCollection = observable(new Collection<ITask>(firestore, getCollection.bind(this, "clients"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));

        this.teamsCollection = observable(new Collection<ITeam>(firestore, getCollection.bind(this, "teams"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));
    }

    @computed
    public get clients() {
        return Array.from(this.clientsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id }));
    }

    @computed
    public get teams() {
        return Array.from(this.teamsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id, isSelected: doc.id === this.teamId }));
    }
}