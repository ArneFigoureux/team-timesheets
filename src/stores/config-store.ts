import { observable, computed } from 'mobx';
import { Collection, ICollection, RealtimeMode, FetchMode } from "firestorable";
import { IRootStore } from './root-store';
import { firestore } from '../firebase/my-firebase';
import { IProject, ITask, IClient, IClientData, ITeam, ITeamData, ITaskData, IConfig, ConfigValue } from '../../common/dist';

import * as serializer from '../../common/serialization/serializer';
import * as deserializer from '../../common/serialization/deserializer';
import { LoginProvider } from '../firebase/types';

export type Configs = {
    // authClientId: string;
    loginProviders: LoginProvider[];
}

export interface IConfigStore extends ConfigStore {};

export class ConfigStore implements IConfigStore {
    //private readonly _rootStore: IRootStore;
    readonly tasks: ICollection<IProject, ITaskData>;
    readonly clientsCollection: ICollection<IClient>;
    readonly teamsCollection: ICollection<ITeam, ITeamData>;
    readonly configsCollection: Collection<IConfig>;

    @observable.ref taskId?: string;
    @observable.ref clientId?: string;
    @observable.ref teamId?: string;

    constructor(_rootStore: IRootStore) {
        // this._rootStore = rootStore;
        const deps = { logger: console.log };

        this.tasks = new Collection<ITask, ITaskData>(
            firestore,
            "tasks",
            {
                realtimeMode: RealtimeMode.on,
                fetchMode: FetchMode.once,
                query: ref => ref.orderBy("name_insensitive"),
                serialize: serializer.convertNameWithIcon,
                deserialize: deserializer.convertNameWithIcon,
            },
            deps,
        );

        this.clientsCollection =
            new Collection<IClient, IClientData>(
                firestore,
                "clients",
                {
                    realtimeMode: RealtimeMode.on,
                    fetchMode: FetchMode.once,
                    query: ref => ref.orderBy("name_insensitive"),
                    serialize: serializer.convertNameWithIcon,
                    deserialize: deserializer.convertNameWithIcon,
                },
                deps,
            );

        this.teamsCollection = new Collection<ITeam, ITeamData>(
            firestore,
            "teams",
            {
                realtimeMode: RealtimeMode.on,
                fetchMode: FetchMode.once,
                query: ref => ref.orderBy("name_insensitive"),
                serialize: serializer.convertNameWithIcon,
                deserialize: deserializer.convertNameWithIcon,
            },
            deps,
        );

        this.configsCollection = new Collection<IConfig>(
            firestore,
            "configs",
            {
                fetchMode: FetchMode.once,
            }
        )
    }

    @computed
    public get clients() {
        return Array.from(this.clientsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id }));
    }

    @computed
    public get teams() {
        return Array.from(this.teamsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id, isSelected: doc.id === this.teamId }));
    }

    // To investigate: does getConfigValue needs mobx @computed attribute?
    public getConfigValue<T>(key: string): T;
    public getConfigValue<T>(key: string, isRequired: false): T | undefined;
    public getConfigValue<T extends ConfigValue>(key: string, isRequired = true): T | undefined {
        const doc = this.configsCollection.docs.find(c => c.data!.key === key);

        if (!doc) {
            if (isRequired) throw new Error(`Required config '${key}' is missing.`);
            else return undefined;
        }

        return doc.data!.value as unknown as T;
    }
}
