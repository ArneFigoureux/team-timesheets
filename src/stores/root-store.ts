import { IRegistrationsStore, RegistrationStore } from "./registration-store";
import { IConfigStore, ConfigStore } from "./config-store";
import { IUserStore, UserStore } from "./user-store";
import { IViewStore, ViewStore } from "./view-store";
import { RouterStore } from "mobx-router";
import { IReportStore, ReportStore } from "./report-store";
import { DashboardStore, IDashboardStore } from "./dashboard-store";
import { firestore } from "../firebase/my-firebase";
import { IProjectStore, ProjectStore } from "./project-store";

export interface IRootStore {

    readonly user: IUserStore;
    readonly view: IViewStore;
    readonly router: RouterStore;
    readonly timesheets: IRegistrationsStore;
    readonly reports: IReportStore;
    readonly config: IConfigStore;
    readonly projects: IProjectStore;
    readonly getCollection: (name: string) => firebase.firestore.CollectionReference;
}

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router = new RouterStore();
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: IProjectStore;

    public readonly getCollection: (name: string) => firebase.firestore.CollectionReference;

    constructor(getCollection: (name: string) => firebase.firestore.CollectionReference) {
        this.getCollection = getCollection;

        this.user = new UserStore(this);
        this.view = new ViewStore(this);
        this.config = new ConfigStore(this, getCollection);
        this.timesheets = new RegistrationStore(this);
        this.reports = new ReportStore(this);
        this.dashboard = new DashboardStore(this);
        this.projects = new ProjectStore(this);

    }
};

const store = (window as any)["store"] = new Store(name => firestore.collection(name));

export default store;

