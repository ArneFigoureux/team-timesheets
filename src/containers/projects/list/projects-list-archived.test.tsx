import React from "react";
import { initTestFirestore, deleteFirebaseAppsAsync } from "../../../__tests__/utils/firebase";
import { Store } from "../../../stores/root-store";
import { TestCollection } from "../../../__tests__/utils/firestorable/collection";
import { IProjectData } from "../../../../common";
import { ArchivedProjectList } from ".";
import { render, waitFor, fireEvent } from "@testing-library/react";
import { canManageProjects } from "../../../rules";
import { act } from "react-dom/test-utils";
const {
    firestore,
    clearFirestoreDataAsync,
    refs: [
        userRef,
        projectRef,
    ]
} = initTestFirestore("project-list-active-test",
    [
        "users",
        "projects",
    ]);

const userCollection = new TestCollection(
    firestore,
    userRef,
);

const projectsCollection = new TestCollection<IProjectData>(firestore, projectRef);
const store = new Store({
    firestore,
});

const setupAsync = () => {
    return Promise.all([
        userCollection.addAsync(
            {
                name: "user 1",
                team: "team-1",
                roles: {
                    user: true,
                }
            },
            "user-1",
        ),
    ]).then(() => {
        store.auth.setUser({
            uid: "user-1",
        } as firebase.User);
    });
};

jest.mock("../../../contexts/store-context", () => ({
    useStore: () => store,
}));

jest.mock("../../../rules");

beforeAll(() => Promise.all([
    clearFirestoreDataAsync(),
    setupAsync(),
]));

afterAll(() => {
    store.dispose();
    return Promise.all([
        deleteFirebaseAppsAsync(),
    ])
});


describe("ProjectListArchived", () => {

    it("renders without projects", () => {
        const { asFragment } = render(<ArchivedProjectList />);

        expect(asFragment()).toMatchSnapshot();
    });

    it("displays archived projects", async () => {
        const projectIds = await projectsCollection.addAsync([
            {
                name: "Project 1",
                name_insensitive: "PROJECT 1",
                icon: "people",
                createdBy: "user-1",
                isArchived: true,
            },
        ]);

        const { getByText, queryByText } = render(<ArchivedProjectList />);
        await waitFor(() => getByText("Project 1"));

        await projectsCollection.updateAsync({
            isArchived: false,
        }, ...projectIds);

        await waitFor(() => expect(queryByText("Project 1")).toBeNull());

        await projectsCollection.updateAsync({
            isArchived: true,
        }, ...projectIds);

        await waitFor(() => getByText("Project 1"));

        await projectsCollection.deleteAsync(...projectIds);

        await waitFor(() => expect(queryByText("Project 1")).toBeNull());
    });

    it("allow to select multiple projects", async () => {
        const projectIds = await projectsCollection.addAsync([
            {
                name: "Project 1",
                name_insensitive: "PROJECT 1",
                icon: "people",
                createdBy: "user-1",
                isArchived: true,
            },
            {
                name: "Project 2",
                name_insensitive: "PROJECT 2",
                icon: "favorite",
                createdBy: "user-1",
                isArchived: true,
            },
        ]);

        const { container } = render(<ArchivedProjectList />);
        const items = container.querySelectorAll(".settings-list-item");
        expect(items.length).toBe(2);

        const checkboxes = container.querySelectorAll("input[type=checkbox]");
        expect(checkboxes.length).toBe(2);

        act(() => {
            fireEvent.click(checkboxes[1]);
            fireEvent.click(items[0]);
        });

        expect(store.view.selection.size).toBe(2);

        // uncheck an item
        act(() => {
            fireEvent.click(checkboxes[0]);
        });

        expect(store.view.selection.size).toBe(1);

        store.view.selection.clear();
        await projectsCollection.deleteAsync(...projectIds);

        await waitFor(() => expect(store.projects.activeProjects.length).toBe(0));
    });

    it("does not allow unauthorised users to click a project", async () => {
        (canManageProjects as any)
            .mockReturnValueOnce(false);

        const projectIds = await projectsCollection.addAsync([
            {
                name: "Project 1",
                name_insensitive: "PROJECT 1",
                icon: "people",
                createdBy: "user-1",
                isArchived: true,
            },
        ]);

        const { container } = render(<ArchivedProjectList />);
        const item = container.querySelector(".settings-list-item");
        expect(item).not.toBeNull();

        act(() => {
            fireEvent.click(item!);
        });

        await waitFor(() => expect(store.view.selection.size).toBe(0));

        await projectsCollection.deleteAsync(...projectIds);

        await waitFor(() => expect(store.projects.activeProjects.length).toBe(0));

    });
});
