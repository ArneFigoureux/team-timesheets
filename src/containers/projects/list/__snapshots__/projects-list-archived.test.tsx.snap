// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectListArchived renders without projects 1`] = `
<DocumentFragment>
  <div
    class="settings-list"
  />
</DocumentFragment>
`;
